### PR TITLE
fix(lsp): create missing directory before creating file

### PR DESCRIPTION
**Pull request by shaunsingh**
Rust-analyzer has a "create module" assist, where it can create a module for you 

e.x. Hovering over the <nameofmodule> 

`mod <nameofmodule>;`

Will prompt you for the following as a code action:

```
Code actions:
1: Create module at `hello.rs`
2: Create module at `hello/mod.rs`
```

The first one would work fine, but #2 would fail with the following error: 
```
Error executing vim.schedule lua callback: /usr/share/nvim/runtime/lua/vim/lsp/util.lua:
759: attempt to index local 'file' (a nil value)
stack traceback:
        /usr/share/nvim/runtime/lua/vim/lsp/util.lua:759: in function 'create_file'
        /usr/share/nvim/runtime/lua/vim/lsp/util.lua:802: in function 'apply_workspace_edit'
        /usr/share/nvim/runtime/lua/vim/lsp/buf.lua:754: in function 'apply_action'
        /usr/share/nvim/runtime/lua/vim/lsp/buf.lua:808: in function 'on_choice'
        /usr/share/nvim/runtime/lua/vim/ui.lua:54: in function 'select'
        /usr/share/nvim/runtime/lua/vim/lsp/buf.lua:819: in function 'on_code_action_results'
        /usr/share/nvim/runtime/lua/vim/lsp/buf.lua:837: in function 'callback'
        /usr/share/nvim/runtime/lua/vim/lsp.lua:1938: in function 'handler'
        /usr/share/nvim/runtime/lua/vim/lsp.lua:1371: in function ''
        vim/_editor.lua: in function <vim/_editor.lua:0>
```

The issue is `io.open` didn't create the `hello/` directory before attempting to create `mod.rs`. This addition checks and creates a directory if it's missing before calling `io.open`

I'm going off of the precedent of https://github.com/kak-lsp/kak-lsp/commit/24fabbf60a511aa73b9fb343a0a6018eff7f73ac that this should be handled within neovim's lsp client.

### CODE REVIEW

File: runtime/lua/vim/lsp/util.lua [hunk 1]
```diff
@@ -755,7 +755,14 @@ local function create_file(change)
...
   if not opts.ignoreIfExists or opts.overwrite then
-    local file = io.open(fname, 'w')
+    io.create_missing_dir_open = function(filename, mode)
+        local dir = vim.fs.dirname(filename)
+        if vim.fn.isdirectory(dir) == 0 then
+            vim.fn.mkdir(dir, "p")
+        end
+        return io.open(filename, mode).
+    end
+    local file = io.create_missing_dir_open(fname, 'w')
```
Comment (mfussenegger) [inline runtime/lua/vim/lsp/util.lua:765]:
```suggestion
    vim.fn.mkdir(vim.fs.dirname(fname), "p")
    local file = io.create_missing_dir_open(fname, 'w')
```

I think this should do the trick too.
No need to create a local function if it's only called once and in the same place.

The `io` module shouldn't be monkey patched.
`mkdir` with `p` exits silently if the directory already exists, so the up-front check isn't necessary either.

**Review by mfussenegger — COMMENTED** [review]
Left a suggestion and a test-case would be nice

**Comment by shaunsingh** [comment]
Thanks, I'll add tests

**Comment by mfussenegger** [comment]
You can have a look at the existing test-cases here: 

https://github.com/neovim/neovim/blob/6cc6e11929ad76a2dc5204aed95cb9ed1dafde23/test/functional/plugin/lsp_spec.lua#L1888-L1921

You can probably copy & adjust the first test case there to include a child folder

**Comment by mfussenegger** [comment]
Added the test case and merged. Thanks for the PR and pointing out the problem

**Pull request merged**